Scripts/SL-ChartParser.lua

```diff
@@ -403,3 +403,343 @@ ParseChartInfo = function(steps, pn)
+	local WasLastStreamFlipped = false
+	local LastStep -- Option<LDUR>
+	local LastRepeatedFoot -- Option<LDUR>
+	-- TODO: microoptimize(?) by counting `NumLRCrossed` explicitly here,
+	-- and maybe even eg `NumConsecutiveFlipped`. test on longass trancemania songs
+	local StepsLR = {}
+	local AnyStepsSinceLastCommitStream = false
+
+	-- bracket jump stuff
+	-- tracks the last arrow(s) (can be plural in the case of brackets themselves)
+	-- that each of the L and R feet was last on, respectively; however,
+	-- note that this can be flipped while recording a (unflipped?) stream...
+	local LastArrowL = "X"
+	local LastArrowR = "X"
+	-- ...so these track the true state of what happened (which become known only
+	-- after deciding whether to flip, hence are updated every CommitStream)
+	-- NB: these are "X" not "" b/c we use them in `match()`, and "" shouldn't match
+	local TrueLastArrowL = "X"
+	local TrueLastArrowR = "X"
+	local TrueLastFoot = nil
+	local JustBracketed = false
+
+	-- TODO(bracket) - figure out what corner cases this is needed for...?
+	-- local justBracketed = false -- used for tiebreaks
+
+	function CommitStream(tieBreakFoot)
+		local ns = #StepsLR
+		local nx = 0
+		for _i, step in ipairs(StepsLR) do
+			-- count crossed-over steps given initial footing
+			if not step then nx = nx + 1 end
+		end
+
+		local needFlip = false
+		if nx * 2 > ns then
+			-- easy case - more than half the L/R steps in this stream were crossed over,
+			-- so we guessed the initial footing wrong and need to flip the stream.
+			needFlip = true
+		elseif nx * 2 == ns then
+			-- exactly half crossed over. note that flipping the stream will introduce
+			-- a jack (i.e. break the alternating-feet assumption) on the first note,
+			-- whereas leaving it as is will footswitch that note. break the tie by
+			-- looking at history to see if the chart is already more jacky or switchy.
+			-- (otoh, the reverse applies if the previous stream was also flipped.)
+			if tieBreakFoot then
+				-- but first, as a higher priority tiebreaker: if this stream is followed
+				-- by a bracketable jump, choose whichever flipness lets us bracket it.
+				if JustBracketed then
+					-- (however, don't get too overzealous -- if also preceded by a
+					-- bracket jump, that forces the footing, so we can't bracket both)
+					needFlip = false
+				elseif LastFoot then
+					needFlip = (tieBreakFoot == "R")
+				else
+					needFlip = (tieBreakFoot == "L")
+				end
+			elseif NumFootswitches > NumJacks then
+				needFlip = LastFlip -- match flipness of last chunk -> footswitch
+			else
+				needFlip = not LastFlip -- don't match -> jack
+			end
+		end
+
+		-- now that we know the correct flip, see if the stream needs split.
+		-- if "too much" of the stream is *completely* crossed over, force a
+		-- double-step there by splitting the stream to stay facing forward.
+		-- heuristic value (9) chosen by inspection on Subluminal - After Hours.
+		local splitIndex -- Option<int>
```

Comment:
You may already be planning this (hard to tell from the comments below), but you should be able to do this same tracking as you're iterating outside of `CommitStream` and avoid the possible n^2 behavior.